cql3: grammar: reject intValue with no contents

avikivity opened the pull request:
The grammar mistakenly allows nothing to be parsed as an intValue (itself accepted in LIMIT and similar clauses).

Easily fixed by removing the empty alternative. A unit test is added.

Fixes #14705.

cvybhu commented:
I see that we have the same error in `jsonValue`.

```cql
cqlsh:tks> insert into tab JSON;
NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.0.0.1:9042 datacenter1>: ConnectionShutdown('Connection to 127.0.0.1:9042 was closed')})
```
It even crashes Scylla.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2491/

avikivity commented:
@nyh please review

nyh commented on test/cql-pytest/test_bad_grammar.py:
```diff
@@ -0,0 +1,42 @@
+
+
+# A collection of tests for grammar that should be rejected, but wasn't at
+# some point in the past.
```
‎These two lines are a general summary of this test file, which repeats the general summary you have above between the "######" lines. Please pick one version and remove the other one.

avikivity replied:
done

nyh commented on test/cql-pytest/test_bad_grammar.py:
```diff
@@ -0,0 +1,42 @@
+# some point in the past.
+
+# table1 has some pre-set data which the tests below SELECT on (the tests
+# shouldn't write to it).
```
You don't need the data in this test, it's completely irrelevant. So please remove this comment, and all the prepare and execute statements in the table1 fixture. Only need:
```python
@pytest.fixture(scope="module")
def table1(cql, test_keyspace):
    with new_test_table(cql, test_keyspace, "p int, c1 int, c2 int, v int, PRIMARY KEY (p, c1, c2)") as table:
        yield table
```

nyh commented on test/cql-pytest/test_bad_grammar.py:
```diff
@@ -0,0 +1,42 @@
+        # from messing with it accidentally.
+        yield table
+
+### GROUP BY without aggregation:
```
I don't know what this comment refers to. Copy-pasta?

avikivity replied:
removed

avikivity commented:
v2:
 - sanitized the unit test:
    - eliminated duplication in the introductory comment
    - removed redundant work in the fixure
    - removed irrelevant comment

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2588/

nyh reviewed (approved):
Looks good.

Nitpic: I just realized that the new test doesn't need to create a table at all - "SELECT * FROM xyz LIMIT" should be a syntax error even if the table xyz doesn't exist - but I'm fine with the extra useless table, maybe it will be useful in another tests (although I guess not - syntax error is always detected before checking the existence of tables.
But maybe it's more "proper" to have a real table. So I'll commit as-is.

nyh commented:
Unfortunately this PR doesn't cleanly apply to next after your recent "cql3: expr: remove the default constructor" also changed Cql.g. Please wait until next is promoted and rebase (or rebase now on next if you're in a hurry).

avikivity commented:
v3: rebased

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2615/

nyh commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2615/

CI failed on a dtest. The failure seems unrelated to the patch. I couldn't find a similar failure in the past so opened https://github.com/scylladb/scylladb/issues/14776, and restarted the CI.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2617/
